### ISSUE TITLE
[distributed] When one of minio server is not started, minio service is unavailable.

### ISSUE THREAD

**Issue by balamurugana** [minio/minio#2577]
Minio version:

```
Version: 2016-08-30T11:15:07Z
Release-Tag: DEVELOPMENT.2016-08-30T11-15-07Z
Commit-ID: fa82d7ba9f3e1b3f1b88b6bc19034ba76ac969e4
```

Reproducer:
1. start minio servers except one node eg `MINIO_CACHE_SIZE=0 ./minio server m1:/mnt m2:/mnt m3:/mnt m4:/mnt` here `m4:/mnt` is not started.
2. run `mc ls m1myminio/`.  Now one of minio is crashed.

`mc ls --debug m1testbin/` output:

```
$ mc ls --debug m1testbin/
mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111823Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:26 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: 9IA81OO0F0VTULD7

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  2.569046049s

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111826Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:26 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: SNT3HNVOC4QHXV9N

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  264.851773ms

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111827Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:27 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: C1JI264R1V8ZUTJ8

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  265.082757ms

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111829Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:30 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: GNUCL6FP76O1ROSO

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  265.906979ms

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111836Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:36 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: B7KZRFJ2JD0V09HL

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  268.445796ms

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111846Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:47 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: V9IS62ZQ4IEY0E60

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  266.030708ms

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111847Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:48 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: HR3XXJTO8JE1LDG3

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  268.956732ms

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111849Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:49 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: TR90EG49PUNCXPQV

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  263.831247ms

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111853Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:18:53 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: QAHG40DBUHN20QLK

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  269.374188ms

mc: <DEBUG> GET / HTTP/1.1
Host: 192.168.122.16:9000
User-Agent: Minio (linux; amd64) minio-go/2.0.1 mc/2016-08-30T11:15:27Z
Authorization: AWS4-HMAC-SHA256 Credential=minio/20160830/us-east-1/s3/aws4_request, SignedHeaders=expect;host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Expect: 100-continue
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20160830T111859Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 503 Service Unavailable
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: text/xml; charset=utf-8
Date: Tue, 30 Aug 2016 11:19:00 GMT
Server: Minio/DEVELOPMENT.2016-08-30T11-15-07Z (linux; amd64)
Vary: Origin
X-Amz-Request-Id: PBYNMSU0Z8DFFVH2

105
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>XMinioServerNotInitialized</Code><Message>Server not initialized, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  264.663046ms

mc: <ERROR> Unable to list folder. 503 Service Unavailable
 (1) ls.go:133 cmd.doList(..) Tags: [http://192.168.122.16:9000/]
 (0) client-s3.go:975 cmd.(*s3Client).listInRoutine(..)
 Release-Tag:DEVELOPMENT.2016-08-30T11-15-27Z | Commit:85717527d9d1 | Host:faba | OS:linux | Arch:amd64 | Lang:go1.6.3 | Mem:2.9MB/6.9MB | Heap:2.9MB/3.7MB
```

`m1:/mnt` log:

```
$ MINIO_CACHE_SIZE=0 ./minio server m1:/mnt m2:/mnt m3:/mnt m4:/mnt

Endpoint:  http://192.168.122.16:9000  http://127.0.0.1:9000
AccessKey: minio
SecretKey: minio123
Region:    us-east-1

Browser Access:
   http://192.168.122.16:9000  http://127.0.0.1:9000

Command-line Access: https://docs.minio.io/docs/minio-client-quickstart-guide
   $ mc config host add myminio http://192.168.122.16:9000 minio minio123

Object API (Amazon S3 compatible):
   Go:         https://docs.minio.io/docs/golang-client-quickstart-guide
   Java:       https://docs.minio.io/docs/java-client-quickstart-guide
   Python:     https://docs.minio.io/docs/python-client-quickstart-guide
   JavaScript: https://docs.minio.io/docs/javascript-client-quickstart-guide
^[[B^[[AERRO[0046] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0047] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0048] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0050] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0057] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0067] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0068] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0070] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0074] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
ERRO[0080] Unable to initialize object layer.            cause=dial tcp 192.168.122.9:9000: getsockopt: connection refused
```

`m2:/mnt` and `m3:/mnt` do not have any logs.

Host info:

```
192.168.122.16  m1
192.168.122.107 m2
192.168.122.188 m3
192.168.122.9   m4
```

**Comment by vadmeste** [comment]
@balamurugana, are you using go1.7 ? this is golang bug (https://github.com/golang/go/issues/16208) and will be fixed in 1.7.1

**Comment by balamurugana** [comment]
> @balamurugana, are you using go1.7 ? this is golang bug (golang/go#16208) and will be fixed in 1.7.1

oops! I thought I am using go1.6.3.  Rerunning the test with go1.6.3

**Comment by vadmeste** [comment]
> @balamurugana, are you using go1.7 ? this is golang bug (golang/go#16208) and will be fixed in 1.7.1
> oops! I thought I am using go1.6.3. Rerunning the test with go1.6.3

you can kill one minio server too, that's an interesting use case too :)

**Comment by balamurugana** [comment]
> you can kill one minio server too, that's an interesting use case too :)

Yep.  That will be a different test :smile:

**Comment by harshavardhana** [comment]
This is a correct behavior for one of the server to report 503 when server is not initialized since you don't have quorum.  Is there anything else in particular that you see @balamurugana ?

**Comment by harshavardhana** [comment]
> This is a correct behavior for one of the server to report 503 when server is not initialized since you don't have quorum. Is there anything else in particular that you see @balamurugana ?

Correcting myself, here we do have quorum so it is a bug.

**Comment by balamurugana** [comment]
This bug appears in minio version

```
Version: 2016-08-31T08:00:08Z
Release-Tag: DEVELOPMENT.2016-08-31T08-00-08Z
Commit-ID: 0b7eac61d102ee59c6b69fc92cc584e4c54d59ed
```

**Comment by harshavardhana** [comment]
> This bug appears in minio version

Checking @balamurugana

**Comment by harshavardhana** [comment]
@balamurugana - should be fixed in the latest https://github.com/minio/minio/commit/05a64b7198669a0a5dbe283b6816d86af716279c can you validate again?

**Comment by harshavardhana** [comment]
> @balamurugana - should be fixed in the latest 05a64b7 can you validate again?

Going to close this re-open if reproducible again.
